python/pyarrow/scalar.pxi

```diff
@@ -16,1198 +16,704 @@
+        arr = self.array()
+        return None if arr is None else arr.to_pylist()
 
-        ptr = ap.GetValue(self.index, &length)
-        return cp.PyBytes_FromStringAndSize(<const char*>(ptr), length)
 
-    def as_buffer(self):
-        """
-        Return a view over this value as a Buffer object.
-        """
-        cdef:
-            CBinaryArray* ap = <CBinaryArray*> self.sp_array.get()
-            shared_ptr[CBuffer] buf
+cdef class FixedSizeListScalar(ListScalar):
+    pass
 
-        buf = SliceBuffer(ap.value_data(), ap.value_offset(self.index),
-                          ap.value_length(self.index))
-        return pyarrow_wrap_buffer(buf)
 
+cdef class LargeListScalar(ListScalar):
+    pass
 
-cdef class LargeBinaryValue(ArrayValue):
+
+cdef class StructScalar(Scalar):
     """
-    Concrete class for large variable-sized binary array elements.
+    Concrete class for struct scalars.
     """
 
-    def as_py(self):
+    def __len__(self):
+        cdef CStructScalar* sp = <CStructScalar*> self.wrapped.get()
+        return sp.value.size()
+
+    def __getitem__(self, key):
         """
-        Return this value as a Python bytes object.
+        Return the child value for the given field.
+
+        Parameters
+        ----------
+        index : Union[int, str]
+            Index / position or name of the field.
+
+        Returns
+        -------
+        result : Scalar
         """
         cdef:
-            const uint8_t* ptr
-            int64_t length
-            CLargeBinaryArray* ap = <CLargeBinaryArray*> self.sp_array.get()
+            CFieldRef ref
+            CStructScalar* sp = <CStructScalar*> self.wrapped.get()
+
+        if isinstance(key, (bytes, str)):
+            ref = CFieldRef(<c_string> tobytes(key))
+        elif isinstance(key, int):
+            ref = CFieldRef(<int> key)
+        else:
+            raise TypeError('Expected integer or string index')
 
-        ptr = ap.GetValue(self.index, &length)
-        return cp.PyBytes_FromStringAndSize(<const char*>(ptr), length)
+        try:
+            return Scalar.wrap(GetResultValue(sp.field(ref)))
+        except ArrowInvalid:
+            raise IndexError(key)
 
-    def as_buffer(self):
+    def as_py(self):
         """
-        Return a view over this value as a Buffer object.
+        Return this value as a Python dict.
         """
         cdef:
-            CLargeBinaryArray* ap = <CLargeBinaryArray*> self.sp_array.get()
-            shared_ptr[CBuffer] buf
+            CStructScalar* sp = <CStructScalar*> self.wrapped.get()
+            CStructType* dtype = <CStructType*> sp.type.get()
+            vector[shared_ptr[CField]] fields = dtype.fields()
 
-        buf = SliceBuffer(ap.value_data(), ap.value_offset(self.index),
-                          ap.value_length(self.index))
-        return pyarrow_wrap_buffer(buf)
+        if sp.is_valid:
+            return {frombytes(fields[i].get().name()): Scalar.wrap(sp.value[i])
```

Comment:
Should we directly take the `as_py` from the values in the dict as well?

Reply:
This is also what we seem to do on master

Reply:
Updated, also extended the StructScalar's API to match a python mapping.